enterprise/server/runner/runner.go

```diff
@@ -0,0 +1,247 @@
+	}, nil
+}
+
+func (r *runnerService) lookupAPIKey(ctx context.Context) (string, error) {
+	auth := r.env.GetAuthenticator()
+	if auth == nil {
+		return "", status.FailedPreconditionError("Auth was not configured but is required")
+	}
+	u, err := auth.AuthenticatedUser(ctx)
+	if err != nil {
+		return "", err
+	}
+	if u.GetGroupID() == "" {
+		return "", status.FailedPreconditionError("Authenticated user did not have a group ID")
+	}
+
+	q := query_builder.NewQuery(`SELECT * FROM APIKeys`)
+	q.AddWhereClause("group_id = ?", u.GetGroupID())
+	qStr, qArgs := q.Build()
+	k := &tables.APIKey{}
+	if err := r.env.GetDBHandle().Raw(qStr, qArgs...).Take(&k).Error; err != nil {
+		return "", err
+	}
+	return k.Value, nil
+}
+
+func (r *runnerService) checkPreconditions(req *rnpb.RunRequest) error {
+	if req.GetGitRepo().GetRepoUrl() == "" {
+		return status.InvalidArgumentError("A repo url is required.")
+	}
+	if req.GetBazelCommand() == "" {
+		return status.InvalidArgumentError("A bazel command is required.")
+	}
+	return nil
+}
+
+func (r *runnerService) createAction(ctx context.Context, req *rnpb.RunRequest, invocationID string) (*repb.Digest, error) {
+	cache := r.env.GetCache()
+	if cache == nil {
+		return nil, status.UnavailableError("No cache configured.")
+	}
+	apiKey, err := r.lookupAPIKey(ctx)
+	if err != nil {
+		return nil, err
+	}
+	binaryBlob, err := os.ReadFile(r.runnerBinaryPath)
+	if err != nil {
+		return nil, err
+	}
+	runnerBinDigest, err := cachetools.UploadBlobToCAS(ctx, cache, req.GetInstanceName(), binaryBlob)
+	if err != nil {
+		return nil, err
+	}
+	// Save this to use when constructing the command to run below.
+	runnerName := filepath.Base(r.runnerBinaryPath)
+	dir := &repb.Directory{
+		Files: []*repb.FileNode{{
+			Name:         runnerName,
+			Digest:       runnerBinDigest,
+			IsExecutable: true,
+		}},
+	}
+	inputRootDigest, err := cachetools.UploadProtoToCAS(ctx, cache, req.GetInstanceName(), dir)
+	if err != nil {
+		return nil, err
+	}
+	conf := r.env.GetConfigurator()
+	cmd := &repb.Command{
+		EnvironmentVariables: []*repb.Command_EnvironmentVariable{
+			{Name: "BUILDBUDDY_API_KEY", Value: apiKey},
+			{Name: "REPO_USER", Value: req.GetGitRepo().GetUsername()},
+			{Name: "REPO_TOKEN", Value: req.GetGitRepo().GetAccessToken()},
+		},
+		Arguments: []string{
+			"./" + runnerName,
+			"--bes_backend=" + conf.GetAppEventsAPIURL(),
+			"--bes_results_url=" + conf.GetAppBuildBuddyURL() + "/invocation/",
+			"--commit_sha=" + req.GetRepoState().GetCommitSha(),
+			"--pushed_repo_url=" + req.GetGitRepo().GetRepoUrl(),
+			"--target_repo_url=" + req.GetGitRepo().GetRepoUrl(),
+			"--bazel_sub_command=" + req.GetBazelCommand(),
+			"--pushed_branch=master",
+			"--target_branch=master",
+			"--invocation_id=" + invocationID,
+		},
+		Platform: &repb.Platform{
+			Properties: []*repb.Platform_Property{
+				{Name: "container-image", Value: "docker://gcr.io/flame-public/buildbuddy-ci-runner:v2.2.7"},
```

Comment:
nit: mind putting a reminder at the top of the same const `workflow/service.go` to remember to update this string if that one changes? (A shared library might also make sense but I get that we might not want to bother with that for now)